src/System.Net.Http.WinHttpHandler/src/System/Net/Http/WinHttpHandler.cs

```diff
@@ -0,0 +1,2164 @@
+                        "value",
+                        value,
+                        string.Format(SR.net_http_value_must_be_greater_than, 0));
+                }
+
+                CheckDisposedOrStarted();
+                _maxResponseDrainSize = value;
+            }
+        }
+        #endregion
+
+        protected override void Dispose(bool disposing)
+        {
+            if (disposing && !_disposed)
+            {
+                _disposed = true;
+
+                if (_sessionHandle != null)
+                {
+                    _sessionHandle.DangerousRelease();
+                    SafeInternetHandle.DisposeAndClearHandle(ref _sessionHandle);
+                }
+            }
+
+            base.Dispose(disposing);
+        }
+
+#if HTTP_DLL
+        protected internal override Task<HttpResponseMessage> SendAsync(
+            HttpRequestMessage request,
+            CancellationToken cancellationToken)
+#else
+        protected override Task<HttpResponseMessage> SendAsync(
+            HttpRequestMessage request,
+            CancellationToken cancellationToken)
+#endif
+        {
+            if (request == null)
+            {
+                throw new ArgumentNullException("request", SR.net_http_handler_norequest);
+            }
+
+            // Check for invalid combinations of properties.
+            if (_proxy != null && _windowsProxyUsePolicy != WindowsProxyUsePolicy.UseCustomProxy)
+            {
+                throw new InvalidOperationException(SR.net_http_invalid_proxyusepolicy);
+            }
+
+            if (_windowsProxyUsePolicy == WindowsProxyUsePolicy.UseCustomProxy && _proxy == null)
+            {
+                throw new InvalidOperationException(SR.net_http_invalid_proxy);
+            }
+
+            if (_cookieUsePolicy == CookieUsePolicy.UseSpecifiedCookieContainer &&
+                _cookieContainer == null)
+            {
+                throw new InvalidOperationException(SR.net_http_invalid_cookiecontainer);
+            }
+
+            CheckDisposed();
+
+            SetOperationStarted();
+
+            TaskCompletionSource<HttpResponseMessage> tcs = new TaskCompletionSource<HttpResponseMessage>();
+
+            // Create RequestState object and save current values of handler settings.
+            RequestState state = new RequestState();
+            state.Tcs = tcs;
+            state.CancellationToken = cancellationToken;
+            state.RequestMessage = request;
+            state.Handler = this;
+            state.CheckCertificateRevocationList = _checkCertificateRevocationList;
+            state.ServerCertificateValidationCallback = _serverCertificateValidationCallback;
+            state.WindowsProxyUsePolicy = this.WindowsProxyUsePolicy;
+            state.Proxy = this.Proxy;
+            state.ServerCredentials = _serverCredentials;
+
+            try
+            {
+                Task.Factory.StartNew(StartRequest, state);
```

Comment:
It may not be necessary here, but it also shouldn't hurt, and I included it because it's what you get when you use Task.Run (here you need to use StartNew though in order to pass in a state object).  Tasks have the ability to have a parent-child relationship.  A task can create itself as AttachToParent, in which case the parent task won't complete until the child task completes, and any exceptions from the child task will bubble up to the parent and cause the parent to fault as well.  DenyChildAttach just lets a task say that it doesn't want to let any tasks attach to it, which makes behavior of the task a bit more predictable.  In this case, no one's waiting on your task, though, and you control all of the code being launched in the task, so it's not a big deal either way.